package fetch.cfg.device;

// Notes
// -----
//
// * May be able to eliminate the nchannels field.  Protobuf will generate a channel_size()
//   method that gives the correct value.
// * rename channel to channels

////////////////////////
message NIScopeDigitizer 
{
  message Channel
  { 
    enum Coupling {                                       // NISCOPE_ATTR_VERTICAL_COUPLING. ENUM values conform to NIScope
      AC = 0;
      DC = 1;
      GND = 2;
    }
    optional string   name     = 1 [default = "0"];       // Null terminated string with channel syntax: e.g. "0-2,7"                                                               
    optional double   range    = 2 [default = 1.0];       // Volts peak to peak.  (NI-5105) Can be 0.05, 0.2, 1, or 6 V for 50 Ohm.  Resolution is 0.1% dB.                         
    optional Coupling coupling = 3 [default = DC];        // Specifies how to couple the input signal. Refer to NISCOPE_ATTR_VERTICAL_COUPLING for more information.                
    optional bool     enabled  = 4 [default = false];     // Specifies whether the channel is enabled for acquisition. Refer to NISCOPE_ATTR_CHANNEL_ENABLED for more information.  
  }

  optional string   name        = 1 [default = "Dev3"];   // NI device name: e.g. "Dev6"                                                                 
  optional double   sample_rate = 2 [default = 60000000]; // samples/second                                                                              
  optional uint32   record_size = 3 [default = 7196];     // samples per scan                                                                            
  optional uint32   num_records = 4 [default = 256];      // number of records per acquire call.                                                         
  optional double   reference   = 5 [default = 0.0];      // as a percentage                                                                             
  optional string   chan_names  = 6 [default = "0"];      // the channels to acquire data on (NI Channel String syntax)                                  
  optional uint32   nchannels   = 7 [default = 1];        // number of channels to independently configure - typically set to DIGITIZER_MAX_NUM_CHANNELS 
  repeated Channel  channel     = 8;                      // array of channel configurations                                                             
}

///////////////////////
message AlazarDigitizer
{
  optional string name = 1 [default = "placeholder!"];
}

/////////////////
message Digitizer
{
  enum DigitizerType
  { 
    NIScope   = 0;
    Alazar    = 1;
    Simulated = 2;
  }
  optional DigitizerType  kind =  1 [default = NIScope];

  optional NIScopeDigitizer niscope = 2;
  optional AlazarDigitizer  alazar  = 3;
}